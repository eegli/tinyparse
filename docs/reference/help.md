# Help

> This document describes how **help options** can be configured.

All good CLI apps have some way of providing help to the user. Tinyparse is no different. You can register usage information and tokens that, when present in the input, will trigger the help text to be printed to the console .

You can register which a subcommand and flags trigger the help text with the `.setHelp()` method:

```ts
import { Parser } from '@eegli/tinyparse';

const parser = new Parser()
  .option('foo', {
    longFlag: '--foo',
    shortFlag: '-f',
    required: true,
    defaultValue: '',
    description: 'Foo option',
  })
  .option('bar', {
    longFlag: '--bar',
    defaultValue: new Date(),
    description: 'Foo option',
  })
  .subcommand('baz', {
    args: ['arg'] as const,
    handler: () => {},
    description: 'Baz command',
  })
  .setHelp({
    appName: 'my-cli',
    summary: 'A brief description of my-cli',
    command: 'help',
    flags: ['--help', '-h'],
  })
  .defaultHandler()
  .parse(['help'])
  .call();
```

This will print the following to the console:

```sh
A brief description of my-cli

Usage: my-cli [command] <...flags>

Commands
   baz <arg>
   - Baz command

Required flags
   -f, --foo [string]
   Foo option

Optional flags
   --bar [date]
   Foo option

To view this help message, run "my-cli help" or add --help or -h to any command
```

When you set your help configuration, Tinyparse will validate the arguments to make sure there are no conflicts with existing flags or subcommands. Note that subsequent calls to `.setHelp()` will overwrite the previous configuration.
