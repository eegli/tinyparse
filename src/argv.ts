import { ValidationError } from './error';
import { CommandOptions } from './types';
import Utils from './utils';

export class ArgvTransformer {
  public static transform(
    argv: string[],
  ): [Map<string, string | boolean>, string[]] {
    const flagMap = new Map<string, string | boolean>();

    const positionals: string[] = [];
    let isPositional = true;

    for (let i = 0; i < argv.length; i++) {
      const arg = argv[i];

      // A short flag is also a long flag
      const isShortFlag = Utils.isShortFlag(arg);

      if (!isShortFlag && isPositional) {
        positionals.push(arg);
        // Collect positionals
        continue;
      }

      isPositional = false;

      if (!isShortFlag) continue;

      const splitted = Utils.splitAtFirst(arg, '=');

      const [flag] = splitted;
      let [, flagVal] = splitted;

      if (!flagVal) flagVal = argv[i + 1];

      // Assume boolean flag
      if (!flagVal || Utils.isShortFlag(flagVal)) {
        flagMap.set(flag, true);
        // Assume string or number
      } else {
        flagMap.set(flag, flagVal);
      }
    }

    return [flagMap, positionals];
  }

  public static validateCommands(
    positionals: string[],
    commandOptions: CommandOptions,
  ): void {
    const availableCommands = Object.keys(commandOptions);

    // No commands are specified in the config
    if (availableCommands.length === 0) return;

    const [command, ...args] = positionals;

    // User did not specify a command
    if (!command || !commandOptions[command]) return;

    const { args: expectedPattern } = commandOptions[command];

    // Allow any number of command arguments
    if (typeof expectedPattern === 'string') return;

    // Number of arguments is satisfied
    if (expectedPattern.length === args.length) return;

    const tooFewArgs = expectedPattern.length > args.length;

    if (tooFewArgs) {
      throw new ValidationError(
        `Invalid usage of command '${command}'. Too few arguments`,
      );
    }
    throw new ValidationError(
      `Invalid usage of command '${command}'. Too many arguments`,
    );
  }
}
