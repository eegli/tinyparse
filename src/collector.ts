import { WithPositionalArgs } from './types';

export class Collector<T> {
  constructor(private readonly defaults: T) {}
  public collectWithPositionals(positionals: string[]): WithPositionalArgs<T> {
    return {
      ...this.collect(),
      _: positionals,
    };
  }

  public collect(): T {
    return this.defaults;
  }
}
