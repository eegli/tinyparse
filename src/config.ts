import {
  AnyGlobal,
  CommandOptionsMap,
  DefaultHandler,
  ErrorHandler,
  FlagOptionsMap,
  FlagValueRecord,
  MetaOptions,
  SubparserOptionsMap,
} from './types';

interface CoreConfig<O extends FlagValueRecord, G extends AnyGlobal> {
  meta: MetaOptions;
  options: FlagOptionsMap;
  commands: CommandOptionsMap<O, G>;
  parsers: SubparserOptionsMap;
}

export interface CommonConfig<O extends FlagValueRecord, G extends AnyGlobal>
  extends CoreConfig<O, G> {
  defaultHandler: DefaultHandler<O, G>;
  globalSetter?: (options: O) => G | Promise<G>;
  errorHandler?: ErrorHandler;
}

export interface HelpPrinterConfig<
  O extends FlagValueRecord,
  G extends AnyGlobal,
> extends Partial<CoreConfig<O, G>> {}
