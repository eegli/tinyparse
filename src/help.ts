import { FilePathArg, InternalOptions, SimpleRecord } from './types';

type DisplayHelp = {
  defaultValues: SimpleRecord;
  options?: InternalOptions;
  filePathArg?: FilePathArg;
  title?: string;
};

export const displayHelp = ({
  defaultValues,
  options = new Map(),
  title = 'Usage',
  filePathArg,
}: DisplayHelp): string => {
  // Required properties first
  const opts = [...options.values()].sort((a, b) =>
    a.required === b.required ? 0 : a.required ? -1 : 1
  );

  let str = title;

  if (opts.length > 0) {
    str += '\n\n';
  }

  // Maybe no option is required
  if (opts[0]?.required) {
    str += 'Required\n';
  }

  let optionalFlag = true;
  const tab = '   ';

  opts.forEach(({ name, description, required, shortFlag }, idx) => {
    const isLast = idx === opts.length - 1;

    if (optionalFlag && !required) {
      str += 'Optional\n';
      optionalFlag = false;
    }

    str += `${tab}${shortFlag ? `${shortFlag}, ` : ''}`;
    str += `--${name}`;
    str += ` [${typeof defaultValues[name]}]`;
    str += description ? `\n${tab}` + description : '';
    str += isLast ? '' : '\n\n';
  });

  if (filePathArg) {
    const { longFlag, description } = filePathArg;
    str += `\n\n${tab}${longFlag} [string]\n`;
    str += description ? `${tab}${description}\n` : '';
  }
  return str;
};
