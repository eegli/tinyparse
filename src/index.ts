import { displayHelp } from './help';
import { parseObjectLiteral } from './parse';
import { transformArgv, transformOptions } from './transform';
import {
  ParsingOptions,
  PositionalArgs,
  SimpleRecord,
  WithPositionalArgs,
} from './types';

export { ValidationError } from './error';

/**
 * Parser factory function. Returns a parser that is bound to the
 * given default values and options per key.
 *
 * @export
 */
export function createParser<T extends SimpleRecord>(
  defaultValues: T,
  parsingOptions?: ParsingOptions<T>
) {
  const options = transformOptions(parsingOptions);
  const filePathArg = parsingOptions?.filePathArg;

  function parse(input?: Partial<T>): Promise<T>;
  function parse(input?: string[]): Promise<WithPositionalArgs<T>>;
  function parse(input: Partial<T> | string[] = {}): Promise<T> {
    let positionalArgs: PositionalArgs | undefined;
    if (Array.isArray(input)) {
      [input, positionalArgs] = transformArgv<T>({
        argv: input,
        filePathFlag: filePathArg?.longFlag,
        options,
      });
    }
    return parseObjectLiteral({
      defaultValues,
      input,
      options,
      forwardArgs: positionalArgs && { _: positionalArgs },
    });
  }

  return {
    help: function (title?: string): string {
      return displayHelp({ defaultValues, options, title, filePathArg });
    },
    parse,
  };
}
