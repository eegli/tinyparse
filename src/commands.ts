import { CommonConfig } from './config';
import { Parser } from './parser';
import {
  AnyGlobal,
  CommandArgPattern,
  DefaultHandler,
  DowncastFlag,
  ErrorHandler,
  FlagOptions,
  FlagOptionsExt,
  FlagValue,
  FlagValueRecord,
  MetaOptions,
  Subcommand,
  Subparser,
} from './types/internals';

export class CommandBuilder<Options, Globals> {
  #config: CommonConfig = {
    meta: {},
    options: new Map(),
    commands: new Map(),
    parsers: new Map(),
    defaultHandler: () => {},
  };

  #takenFlags = new Set<string>();
  #takenCommands = new Set<string>();

  #validateCommand = (command: string): void => {
    if (this.#takenCommands.has(command)) {
      throw new Error(`Command "${command}" has been declared twice`);
    }
  };

  #validateFlag = (flag: string): void => {
    if (this.#takenFlags.has(flag)) {
      throw new Error(`Flag "${flag}" has been declared twice`);
    }
  };

  #validateOption = (key: string): void => {
    if (this.#config.options.has(key)) {
      throw new Error(`Option "${key}" has been declared twice`);
    }
  };

  #validateOneOf = <T>(id: string, type: T, ...options: T[]): void => {
    const expectedType = typeof type;
    const wrongType = options.find((opt) => typeof opt !== expectedType);
    if (wrongType) {
      throw new Error(
        `OneOf for option "${id}" contains invalid type ${typeof wrongType}, expected ${expectedType}`,
      );
    }
  };

  #tryRegisterCommandToken = (command?: string) => {
    if (command) {
      this.#validateCommand(command);
      this.#takenCommands.add(command);
    }
  };

  #tryRegisterFlagToken = (flag?: string) => {
    if (flag) {
      this.#validateFlag(flag);
      this.#takenFlags.add(flag);
    }
  };

  /**
   * Add an option (flag)
   */
  option<
    K extends string,
    V extends FlagValue,
    R extends boolean,
    O extends unknown[],
    Choices = O[number],
    Values = R extends true ? Choices : V | Choices,
  >(
    key: K,
    opts: FlagOptionsExt<V, R, Choices>,
  ): CommandBuilder<Options & Record<K, Values>, Globals>;
  option<K extends string, V extends FlagValue, R extends boolean>(
    key: K,
    opts: FlagOptions<V, R>,
  ): CommandBuilder<Options & Record<K, DowncastFlag<V>>, Globals>;
  option<K extends string, V extends FlagValue, R extends boolean>(
    key: K,
    opts: FlagOptions<V, R>,
  ) {
    const { longFlag, shortFlag, oneOf, defaultValue } = opts;

    if (oneOf) {
      this.#validateOneOf(key, defaultValue, ...oneOf);
    }

    this.#validateOption(key);
    this.#config.options.set(key, opts);

    this.#tryRegisterFlagToken(longFlag);
    this.#tryRegisterFlagToken(shortFlag);

    return this as CommandBuilder<
      Options & Record<K, DowncastFlag<V>>,
      Globals
    >;
  }

  /**
   * Add a subcommand
   */
  subcommand<A extends CommandArgPattern>(
    command: string,
    opts: Subcommand<Options, Globals, A>,
  ) {
    this.#tryRegisterCommandToken(command);
    this.#config.commands.set(
      command,
      opts as Subcommand<FlagValueRecord, AnyGlobal, A>,
    );
    return this;
  }

  /**
   * Add a subparser
   */
  subparser(command: string, opts: Subparser) {
    this.#tryRegisterCommandToken(command);
    this.#config.parsers.set(command, opts);
    return this;
  }

  /**
   * Set the globals
   */
  setGlobals<G extends AnyGlobal>(
    setGlobals: (options: Options) => G | Promise<G>,
  ) {
    this.#config.globalSetter = setGlobals as (
      options: FlagValueRecord,
    ) => AnyGlobal | Promise<AnyGlobal>;
    return this as CommandBuilder<Options, Globals & G>;
  }

  /**
   * Set metadata
   */
  setMeta(meta: MetaOptions) {
    this.#tryRegisterCommandToken(meta.help?.command);
    this.#tryRegisterCommandToken(meta.version?.command);

    this.#tryRegisterFlagToken(meta.help?.longFlag);
    this.#tryRegisterFlagToken(meta.help?.shortFlag);
    this.#tryRegisterFlagToken(meta.version?.longFlag);
    this.#tryRegisterFlagToken(meta.version?.shortFlag);

    this.#config.meta = meta;
    return this;
  }

  /**
   * Set the error handler
   */
  onError(handler: ErrorHandler) {
    this.#config.errorHandler = handler;
    return this;
  }

  /**
   * Set the default handler
   */
  defaultHandler(handler?: DefaultHandler<Options, Globals>) {
    return new Parser<Options>({
      ...this.#config,
      defaultHandler:
        (handler as DefaultHandler) || this.#config.defaultHandler,
    });
  }
}
