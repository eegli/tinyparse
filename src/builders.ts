import { transformArgv } from './argv';
import { validateCommandArgs } from './commands';
import { collect } from './flags';
import {
  Subcommand,
  Downcast,
  FlagOptionMap,
  FlagRecord,
  FlagOptions,
  FlagArgValue,
  CommandOptionMap,
  CommandArgPattern,
} from './types';

export class CommandBuilder<F extends FlagRecord = Record<never, never>> {
  #flags: FlagOptionMap = new Map();
  #commands: CommandOptionMap<F> = new Map();

  public subcommand<P extends CommandArgPattern>(
    command: string,
    opts: Subcommand<F, P>,
  ) {
    this.#commands.set(command, opts);
    return this;
  }

  public flag<T extends string, V extends FlagArgValue>(
    flag: T,
    opts: FlagOptions<V>,
  ) {
    this.#flags.set(flag, opts);
    // TODO: Figure out how to make this typecheck properly
    return this as unknown as CommandBuilder<F & Record<T, Downcast<V>>>;
  }

  public build() {
    return new DispatchBuilder<F>(this.#flags, this.#commands);
  }
}

export class DispatchBuilder<F extends FlagRecord> {
  #handler?: (flags: F, positionals: string[]) => void;
  #flags: FlagOptionMap;
  #commands: CommandOptionMap<F>;

  constructor(flags: FlagOptionMap, commands: CommandOptionMap<F>) {
    this.#flags = flags;
    this.#commands = commands;
  }

  public default<T extends F>(
    handler: T extends FlagRecord
      ? (flags: T, positionals: string[]) => void
      : never,
  ) {
    this.#handler = handler;
    return this;
  }

  public parse(argv: string[]): {
    flags: F;
    call: () => void;
  } {
    const [flagMap, positionals] = transformArgv(argv);
    const flags = collect(flagMap, this.#flags) as F;

    const [command, ...args] = positionals;
    const commandOpts = this.#commands.get(command);

    if (command && commandOpts) {
      validateCommandArgs(command, commandOpts, args);
      const call = commandOpts.handler.bind(this, flags, args);
      return {
        flags,
        call,
      };
    }
    if (this.#handler) {
      const call = this.#handler?.bind(this, flags, positionals);
      return {
        flags,
        call,
      };
    }
    return {
      flags,
      call: () => {},
    };
  }
}
