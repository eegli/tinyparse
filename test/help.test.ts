import { CommandBuilder } from '../src/commands';
import { HelpPrinter as Help } from '../src/help';
import {
  CommandOptionsMap,
  FlagOptionsMap,
  HelpOptions,
  VersionOptions,
} from '../src/types/internals';

describe('Helper text', () => {
  const requiredflags: FlagOptionsMap = new Map([
    [
      'flagA',
      {
        longFlag: '--flag-a',
        defaultValue: '',
        required: true,
        description: 'The first flag',
      },
    ],
    [
      'flabB',
      {
        longFlag: '--flag-b',
        defaultValue: new Date(),
        required: true,
      },
    ],
  ]);
  const optionalFlags: FlagOptionsMap = new Map([
    [
      'flagD',
      {
        longFlag: '--flag-d',
        shortFlag: '-d',
        defaultValue: 3000,
        description: 'The fourth flag',
      },
    ],
    [
      'flabC',
      {
        longFlag: '--flag-c',
        shortFlag: '-c',
        defaultValue: false,
      },
    ],
  ]);
  const allFlags = new Map([...requiredflags, ...optionalFlags]);
  const commands: CommandOptionsMap = new Map([
    [
      'serve',
      {
        args: ['path'],
        description: 'Serve a directory',
        handler: () => {},
      },
    ],
    [
      'info',
      {
        args: [],
        handler: () => {},
      },
    ],
    [
      'rm',
      {
        args: '...files',
        description: 'Remove files',
        handler: () => {},
      },
    ],
    [
      'ls',
      {
        args: undefined,
        description: 'List a directory',
        handler: () => {},
      },
    ],
  ]);

  const help: HelpOptions = {
    command: 'help',
    longFlag: '--help',
    shortFlag: '-h',
  };

  const version: VersionOptions = {
    command: 'version',
    version: '1.0.0',
    longFlag: '--version',
    shortFlag: '-v',
  };
  const command = 'my-app';
  const summary = 'This is just a text that can be shown to describe the app';

  test('header formatting', () => {
    expect(new Help({ meta: { summary, command } }).formatHeader())
      .toMatchInlineSnapshot(`
      "This is just a text that can be shown to describe the app

      Usage: my-app [command?] <...flags>"
    `);
    expect(
      new Help({ meta: { command } }).formatHeader(),
    ).toMatchInlineSnapshot(`"Usage: my-app [command?] <...flags>"`);
    expect(new Help({ meta: { summary } }).formatHeader())
      .toMatchInlineSnapshot(`
      "This is just a text that can be shown to describe the app

      Usage:"
    `);
    expect(new Help().formatHeader()).toMatchInlineSnapshot(`"Usage:"`);
  });
  test('commands formatting', () => {
    expect(new Help({ commands }).formatCommands()).toMatchInlineSnapshot(`
      "Commands
         info            
         ls              List a directory
         rm <...files>   Remove files
         serve <path>    Serve a directory"
    `);
  });
  test('required flags formatting', () => {
    const p = new Help({ options: requiredflags });
    expect(p.formatOptions()).toMatchInlineSnapshot(`
      "Required flags
         --flag-a [string]   The first flag
         --flag-b [date]     "
    `);
  });
  test('optional flags formatting', () => {
    expect(
      new Help({
        meta: {
          help: {
            longFlag: '--help',
          },
        },
        options: optionalFlags,
      }).formatOptions(),
    ).toMatchInlineSnapshot(`
      "Optional flags
         -c, --flag-c [boolean]   
         -d, --flag-d [number]    The fourth flag
         --help                   Print this help message"
    `);
  });
  test('required and optional flags formatting', () => {
    expect(
      new Help({
        options: allFlags,
      }).formatOptions(),
    ).toMatchInlineSnapshot(`
      "Required flags
         --flag-a [string]        The first flag
         --flag-b [date]          

      Optional flags
         -c, --flag-c [boolean]   
         -d, --flag-d [number]    The fourth flag"
    `);
  });

  test('no configuration', () => {
    expect(new Help().printUsage()).toMatchInlineSnapshot(`"Usage:"`);
  });
  test('full configuration', () => {
    expect(
      new Help({
        meta: {
          command,
          summary,
          help,
          version,
        },
        options: allFlags,
        parsers: new Map([
          [
            'a-subparser',
            {
              parser: new CommandBuilder().defaultHandler(),
              description: 'Subparser description',
            },
          ],

          [
            'x-subparser',
            {
              parser: new CommandBuilder().defaultHandler(),
            },
          ],
        ]),

        commands,
      }).printUsage(),
    ).toMatchInlineSnapshot(`
      "This is just a text that can be shown to describe the app

      Usage: my-app [command?] <...flags>

      Commands
         a-subparser     Subparser description
         info            
         ls              List a directory
         rm <...files>   Remove files
         serve <path>    Serve a directory
         x-subparser     
         help            Print this help message
         version         Print the version

      Required flags
         --flag-a [string]        The first flag
         --flag-b [date]          

      Optional flags
         -c, --flag-c [boolean]   
         -d, --flag-d [number]    The fourth flag
         -h, --help               Print this help message
         -v, --version            Print the version"
    `);
  });
});
