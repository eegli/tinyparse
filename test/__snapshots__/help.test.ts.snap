// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Helper text commands 1`] = `
"Usage

Required flags
   --nonverbose [boolean]

Optional flags
   --verbose [boolean]"
`;

exports[`Helper text creates helper text with descriptions 1`] = `
"CLI usage

do stuff with this CLI

Required flags
   --color [string]
   A color

   -p, --port [number]
   The port to listen on

Optional flags
   --UserId [string]

   -wa, --withAuth [boolean]
   Require authentication for this action

   -c, --config [string]
   The config file to use
"
`;

exports[`Helper text creates helper text with no descriptions 1`] = `
"Usage

Optional flags
   --id [string]

   --port [number]

   --withAuth [boolean]"
`;

exports[`Helper text file flag helper text only 1`] = `
"Usage

   --config [string]
"
`;

exports[`Helper text file flag helper text only 2`] = `
"Usage

   -c, --config [string]
"
`;

exports[`Helper text with decamelization and custom long flag 1`] = `
"Usage

Optional flags
   --color [string]

   --user-id [string]

   --with-auth [boolean]"
`;
